iothub/service/src/Messaging/MessagingClient.cs

```diff
@@ -0,0 +1,311 @@
+        protected MessagingClient()
+        {
+        }
+
+        internal MessagingClient(
+            string hostName,
+            IotHubConnectionProperties credentialProvider,
+            IotHubServiceClientOptions options)
+        {
+            _hostName = hostName;
+            _credentialProvider = credentialProvider;
+            _clientOptions = options;
+            _connection = new IotHubConnection(credentialProvider, options.UseWebSocketOnly, options.TransportSettings, options);
+            _openTimeout = IotHubConnection.DefaultOpenTimeout;
+            _operationTimeout = IotHubConnection.DefaultOperationTimeout;
+            _faultTolerantSendingLink = new FaultTolerantAmqpObject<SendingAmqpLink>(CreateSendingLinkAsync, _connection.CloseLink);
+            _sendingPath = "/messages/deviceBound";
+        }
+
+        /// <summary>
+        /// Open the MessagingClient instance.
+        /// </summary>
+        public virtual async Task OpenAsync()
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, $"Opening MessagingClient", nameof(OpenAsync));
+            try
+            {
+                using var ctx = new CancellationTokenSource(_openTimeout);
+                await _faultTolerantSendingLink.OpenAsync(ctx.Token).ConfigureAwait(false);
+            }
+            catch(Exception ex)
+            {
+                if (Logging.IsEnabled)
+                    Logging.Error(this, $"{nameof(OpenAsync)} threw an exception: {ex}", nameof(OpenAsync));
+                throw;
+            }
+            finally
+            {
+                if (Logging.IsEnabled)
+                    Logging.Exit(this, $"Opening MessagingClient", nameof(OpenAsync));
+            }
+        }
+
+        /// <summary>
+        /// Close the MessagingClient instance.
+        /// </summary>
+        public virtual async Task CloseAsync()
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, $"Closing MessagingClient", nameof(CloseAsync));
+
+            try
+            {
+                await _faultTolerantSendingLink.CloseAsync().ConfigureAwait(false);
+                await _connection.CloseAsync().ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                if (Logging.IsEnabled)
+                    Logging.Error(this, $"{nameof(CloseAsync)} threw an exception: {ex}", nameof(CloseAsync));
+                throw;
+            }
+            finally
+            {
+                if (Logging.IsEnabled)
+                    Logging.Exit(this, $"Closing MessagingClient", nameof(CloseAsync));
+            }
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Dispose the MessagingClient instance.
+        /// </summary>
+        public virtual void Dispose(bool disposing)
```

Comment:
should be protected

Reply:
https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose